{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyarrow\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "#jupyter nbconvert homer_report.ipynb --no-input --to html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HOMER data integration to WGS of DSD project"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What is HOMER?\n",
    "HOMER is a software for motif discovery and next-gen sequencing analysis.\n",
    "we use the findMotifsGenome.pl program to perform de novo motif discovery as well as check the enrichment of known motifs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Our process:\n",
    "We've set up a nextflow pipeline to run HOMER on our hATAC_mATAC regions to find enriched and de-novo motifs. This is an example output data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PositionID</th>\n",
       "      <th>Offset</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Motif Name</th>\n",
       "      <th>Strand</th>\n",
       "      <th>MotifScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15822_Y35</td>\n",
       "      <td>671</td>\n",
       "      <td>CCTTTGTGGG</td>\n",
       "      <td>Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...</td>\n",
       "      <td>+</td>\n",
       "      <td>7.741960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15822_Y35</td>\n",
       "      <td>685</td>\n",
       "      <td>CCTTTGTGGG</td>\n",
       "      <td>Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...</td>\n",
       "      <td>+</td>\n",
       "      <td>7.741960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PS.SE42022</td>\n",
       "      <td>1845</td>\n",
       "      <td>CCATTGTCCC</td>\n",
       "      <td>Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...</td>\n",
       "      <td>+</td>\n",
       "      <td>8.756874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PS.SE42020</td>\n",
       "      <td>1251</td>\n",
       "      <td>CCTTTGTACC</td>\n",
       "      <td>Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...</td>\n",
       "      <td>+</td>\n",
       "      <td>7.464307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PS.SE42019</td>\n",
       "      <td>239</td>\n",
       "      <td>GAAACAAAGC</td>\n",
       "      <td>Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...</td>\n",
       "      <td>-</td>\n",
       "      <td>8.253506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942841</th>\n",
       "      <td>83480_Y35_GR.PG44</td>\n",
       "      <td>974</td>\n",
       "      <td>ACATCAAAGG</td>\n",
       "      <td>Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer</td>\n",
       "      <td>+</td>\n",
       "      <td>11.264622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942842</th>\n",
       "      <td>83486_XY1235_PG39</td>\n",
       "      <td>161</td>\n",
       "      <td>GCATCAAAGG</td>\n",
       "      <td>Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer</td>\n",
       "      <td>+</td>\n",
       "      <td>9.213214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942843</th>\n",
       "      <td>83486_XY1235_PG39</td>\n",
       "      <td>582</td>\n",
       "      <td>GCTTCAAAGG</td>\n",
       "      <td>Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer</td>\n",
       "      <td>+</td>\n",
       "      <td>9.170481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942844</th>\n",
       "      <td>GR.PG.PS.SE28</td>\n",
       "      <td>1349</td>\n",
       "      <td>TCATCAAAGG</td>\n",
       "      <td>Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer</td>\n",
       "      <td>+</td>\n",
       "      <td>8.886311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942845</th>\n",
       "      <td>83509_Y35</td>\n",
       "      <td>68</td>\n",
       "      <td>ACTTCAAAGG</td>\n",
       "      <td>Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer</td>\n",
       "      <td>+</td>\n",
       "      <td>11.221889</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>942846 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               PositionID  Offset    Sequence  \\\n",
       "0               15822_Y35     671  CCTTTGTGGG   \n",
       "1               15822_Y35     685  CCTTTGTGGG   \n",
       "2              PS.SE42022    1845  CCATTGTCCC   \n",
       "3              PS.SE42020    1251  CCTTTGTACC   \n",
       "4              PS.SE42019     239  GAAACAAAGC   \n",
       "...                   ...     ...         ...   \n",
       "942841  83480_Y35_GR.PG44     974  ACATCAAAGG   \n",
       "942842  83486_XY1235_PG39     161  GCATCAAAGG   \n",
       "942843  83486_XY1235_PG39     582  GCTTCAAAGG   \n",
       "942844      GR.PG.PS.SE28    1349  TCATCAAAGG   \n",
       "942845          83509_Y35      68  ACTTCAAAGG   \n",
       "\n",
       "                                               Motif Name Strand  MotifScore  \n",
       "0       Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...      +    7.741960  \n",
       "1       Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...      +    7.741960  \n",
       "2       Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...      +    8.756874  \n",
       "3       Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...      +    7.464307  \n",
       "4       Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...      -    8.253506  \n",
       "...                                                   ...    ...         ...  \n",
       "942841        Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer      +   11.264622  \n",
       "942842        Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer      +    9.213214  \n",
       "942843        Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer      +    9.170481  \n",
       "942844        Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer      +    8.886311  \n",
       "942845        Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer      +   11.221889  \n",
       "\n",
       "[942846 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homer_output_file = \"../25_12/keyTF_motifs_in_peaks.csv\"\n",
    "homer_output = pd.read_csv(homer_output_file)\n",
    "homer_output.iloc[:, 1:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results columns:\n",
    "PositionID is the name of the ATAC interval, \n",
    "Offset is from the start of the ATAC interval, \n",
    "Sequence is the BS, \n",
    "Motif Name includes the TF name and the source of the experimental data, \n",
    "Strand of the BS,\n",
    "Score.\n",
    "\n",
    "The score calculates the likelihood that a given sequence matches the motif represented by the PWM compared to a background model (often assuming random occurrence of bases). Higher log odds scores indicate a better match or stronger similarity between the motif and the input sequence. Homer recommands a threshold around 5-10, and 5 is used as a default. \n",
    "*We currently don't use a higher threshold*\n",
    "This is the distribution of the scores:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    942846.000000\n",
      "mean          7.560098\n",
      "std           1.316626\n",
      "min           5.195305\n",
      "25%           6.634124\n",
      "50%           7.574695\n",
      "75%           8.370506\n",
      "max          12.770999\n",
      "Name: MotifScore, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "summary_stats = homer_output['MotifScore'].describe()\n",
    "print(summary_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As part of the pipeline, we filter the results to keep only data about our TF of interest. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Out of our list of TF of interest, not all are in HOMER:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>exists_in_homer_data</th>\n",
       "      <th>source</th>\n",
       "      <th>origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GATA4</td>\n",
       "      <td>True</td>\n",
       "      <td>Gata4(Zf)/Heart-Gata4-ChIP-Seq(GSE35151)/Homer</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LHX9</td>\n",
       "      <td>True</td>\n",
       "      <td>LHX9(Homeobox)/Hct116-LHX9.V5-ChIP-Seq(GSE1168...</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR5A1</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AR</td>\n",
       "      <td>True</td>\n",
       "      <td>FoxL2(Forkhead)/Ovary-FoxL2-ChIP-Seq(GSE60858)...</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DMRT1</td>\n",
       "      <td>True</td>\n",
       "      <td>DMRT1(DM)/Testis-DMRT1-ChIP-Seq(GSE64892)/Homer</td>\n",
       "      <td>mouse + human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SOX8</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SOX9</td>\n",
       "      <td>True</td>\n",
       "      <td>Sox9(HMG)/Limb-SOX9-ChIP-Seq(GSE73225)/Homer</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SOX10</td>\n",
       "      <td>True</td>\n",
       "      <td>Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...</td>\n",
       "      <td>rat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRY</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ESR2</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FOXL2</td>\n",
       "      <td>True</td>\n",
       "      <td>FoxL2(Forkhead)/Ovary-FoxL2-ChIP-Seq(GSE60858)...</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>LEF1</td>\n",
       "      <td>True</td>\n",
       "      <td>LEF1(HMG)/H1-LEF1-ChIP-Seq(GSE64758)/Homer</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RUNX1</td>\n",
       "      <td>True</td>\n",
       "      <td>RUNX1(Runt)/Jurkat-RUNX1-ChIP-Seq(GSE29180)/Homer</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TCF3</td>\n",
       "      <td>True</td>\n",
       "      <td>Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TCF12</td>\n",
       "      <td>True</td>\n",
       "      <td>Tcf12(bHLH)/GM12878-Tcf12-ChIP-Seq(GSE32465)/H...</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>WT1</td>\n",
       "      <td>True</td>\n",
       "      <td>WT1(Zf)/Kidney-WT1-ChIP-Seq(GSE90016)/Homer</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF  exists_in_homer_data  \\\n",
       "0   GATA4                  True   \n",
       "1    LHX9                  True   \n",
       "2   NR5A1                 False   \n",
       "3      AR                  True   \n",
       "4   DMRT1                  True   \n",
       "5    SOX8                 False   \n",
       "6    SOX9                  True   \n",
       "7   SOX10                  True   \n",
       "8     SRY                 False   \n",
       "9    ESR1                 False   \n",
       "10   ESR2                 False   \n",
       "11  FOXL2                  True   \n",
       "12   LEF1                  True   \n",
       "13  RUNX1                  True   \n",
       "14   TCF3                  True   \n",
       "15  TCF12                  True   \n",
       "16    WT1                  True   \n",
       "\n",
       "                                               source         origin  \n",
       "0      Gata4(Zf)/Heart-Gata4-ChIP-Seq(GSE35151)/Homer          mouse  \n",
       "1   LHX9(Homeobox)/Hct116-LHX9.V5-ChIP-Seq(GSE1168...          human  \n",
       "2                                                None           None  \n",
       "3   FoxL2(Forkhead)/Ovary-FoxL2-ChIP-Seq(GSE60858)...          mouse  \n",
       "4     DMRT1(DM)/Testis-DMRT1-ChIP-Seq(GSE64892)/Homer  mouse + human  \n",
       "5                                                None           None  \n",
       "6        Sox9(HMG)/Limb-SOX9-ChIP-Seq(GSE73225)/Homer          mouse  \n",
       "7   Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...            rat  \n",
       "8                                                None           None  \n",
       "9                                                None           None  \n",
       "10                                               None           None  \n",
       "11  FoxL2(Forkhead)/Ovary-FoxL2-ChIP-Seq(GSE60858)...          mouse  \n",
       "12         LEF1(HMG)/H1-LEF1-ChIP-Seq(GSE64758)/Homer          human  \n",
       "13  RUNX1(Runt)/Jurkat-RUNX1-ChIP-Seq(GSE29180)/Homer          human  \n",
       "14        Tcf3(HMG)/mES-Tcf3-ChIP-Seq(GSE11724)/Homer          mouse  \n",
       "15  Tcf12(bHLH)/GM12878-Tcf12-ChIP-Seq(GSE32465)/H...          human  \n",
       "16        WT1(Zf)/Kidney-WT1-ChIP-Seq(GSE90016)/Homer          mouse  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_list = pd.read_csv(\"../../useful_files/tf_list.csv\", header=None)\n",
    "tf_list.columns = ['TF']\n",
    "tf_list['exists_in_homer_data'] = tf_list['TF'].apply(lambda x: any(homer_output['Motif Name'].str.contains(x, case=False)))\n",
    "tf_list['source'] = tf_list['TF'].apply(lambda x: next((source for source in homer_output['Motif Name'] if x.lower() in source.lower()), None))\n",
    "tf_list['origin'] = ['mouse', 'human',  None, 'mouse', 'mouse + human', None, 'mouse', 'rat', None, None, None, 'mouse', 'human', 'human', 'mouse', 'human', 'mouse']\n",
    "tf_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132)/Homer'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_list['source'][7]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So instead of NR5A1 we take NR5A2, but the rest of the 'FALSE' don't have a replacement."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Eventually we save the parsed data to \"homer_TFBS_results.prq\" from which we extract relevant data according to genomic coordinates using \"extractTFBS.R\"."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
